### PR TITLE
Native cross-compilation with goreleaser-xx and CI enhancements

### PR DESCRIPTION
**Pull request by crazy-max**
Carry #87

@jzelinskie First of all thank you for this great project! Really useful but I think the CI needs some love so as a modest gift, here is a PR that offers some enhancements for the current workflow:

* Use [buildx bake](https://github.com/docker/buildx) and [goreleaser-xx](https://github.com/crazy-max/goreleaser-xx) (with also [xx project](https://github.com/tonistiigi/xx/)) for native cross-compilation in the Dockerfile.
* Create artifacts with `artifact` (current platform) or `artifact-all` target for cross-compilation:
  * `linux/amd64`
  * `linux/arm/v6`
  * `linux/arm/v7`
  * `linux/arm64`
  * `linux/ppc64le`
  * `linux/riscv64`
  * `linux/s390x`
* Sandboxed test and lint with bake targets
* Mutualize tests and handle them through bake and GitHub Actions with coverage
* USe codecov/codecov-action to upload coverage results
* Support Docker multi-platform image
* GitHub Actions
  * On git `push tag` event semver like `v0.0.8`:
    * Artifacts will be available in the pipeline through [`actions/upload-artifact`](https://github.com/actions/upload-artifact/)
    * Will create the GitHub Release and push artifacts
    * Will create and push Docker tags `ghcr.io/jzelinskie/faq:0.0.8` / `ghcr.io/jzelinskie/faq:latest`
  * On git `push` (main branch)
    * Artifacts will be available in the pipeline through [`actions/upload-artifact`](https://github.com/actions/upload-artifact/)
    * Will push Docker tag `ghcr.io/jzelinskie/faq:edge`
  * On git `pull_request` event
    * Artifacts will be available in the pipeline through [`actions/upload-artifact`](https://github.com/actions/upload-artifact/)

Everything is already in place and tested on https://github.com/crazy-max/faq as well as the [GitHub Actions pipeline](https://github.com/crazy-max/faq/actions) if you want to take a look.

image
> https://github.com/crazy-max/faq/releases/tag/0.0.8

image
> https://github.com/crazy-max/faq/pkgs/container/faq

image
> Codecov https://app.codecov.io/gh/crazy-max/faq

Here is how to build with buildx:

```shell
# create builder for multi-arch build and cache feature (see below)
docker buildx create --name faq-builder

# build docker image and output to docker with faq:local tag (default)
docker buildx bake image-local

# build multi-platform image
docker buildx bake image-all

# create the artifact matching your current platform in ./dist
docker buildx bake artifact

# create artifacts for many platforms in ./dist
docker buildx bake artifact-all
```

For now windows and darwin cross-compilation are skipped but we can do that in a follow-up.

Let me know if you have any questions.

### CODE REVIEW

**Comment by jzelinskie** [comment]
Thanks! There are a bunch of new tricks that I've never seen before! I'm going to take a look in the next few days hopefully when I get some time.

**Comment by crazy-max** [comment]
@jzelinskie Sure let me know!

About the failed jobs looks like a partial disruption on Docker Hub:

```
error: Error response from daemon: Head "https://registry-1.docker.io/v2/moby/buildkit/manifests/buildx-stable-1": received unexpected HTTP status: 503 Service Unavailable
```

You can trigger the build again and it should work.

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -1,83 +1,151 @@
...
+        uses: docker/setup-buildx-action@v1
+      -
+        name: Login to GHCR
+        if: github.event_name != 'pull_request'
         uses: docker/login-action@v1
-        if: github.event_name == 'push'
         with:
-          registry: quay.io
-          username: jzelinskie+faq_gha
-          password: ${{ secrets.QUAY_ROBOT_TOKEN }}
```
Comment (crazy-max) [inline .github/workflows/build.yaml:76]: Oh btw, do you still want to push on Quay? Last time I checked, this registry didn't support multi-platform images.

Comment (jzelinskie) [reply .github/workflows/build.yaml:76]: Quay does support multi-platform images. I looked and there is regular traffic pulling from there, so ideally we don't break those users.

**Comment by jzelinskie** [comment]
I can likely take a look at this PR tomorrow. I think my only concern is the lack of macOS amd64/arm64 builds for releases -- which is what makes everything oh so complicated, haha.

File: .github/workflows/build.yaml (hunk 1 above)
Comment (crazy-max) [reply .github/workflows/build.yaml:76]: Ok will bring back Quay!

**Comment by crazy-max** [comment]
> I can likely take a look at this PR tomorrow. I think my only concern is the lack of macOS amd64/arm64 builds for releases -- which is what makes everything oh so complicated, haha.
>
> We do manage to do this at work on [SpiceDB's goreleaser config](https://github.com/authzed/spicedb/blob/main/.goreleaser.yml).

I have a draft for this one but it requires MacOS SDK and on SpiceDB project it works because CGO is disabled.

**Comment by jzelinskie** [comment]
>What we could do is having a dedicated job to generate win/mac artifacts. Will update this PR with this.

This is what we do for our work projects that do require CGO: take a look at our CLI tool zed's workflow:
https://github.com/authzed/zed/blob/main/.github/workflows/release.yaml

**Comment by crazy-max** [comment]
@jzelinskie Ok going a step further, we are now able to build against `darwin/amd64` with [`goxx`](https://github.com/crazy-max/goxx):
```
#14 1.434    • snapshotting
#14 1.434       • building snapshot...      version=0.0.7-next
#14 1.434    • checking distribution directory
#14 1.434    • loading go mod information
#14 1.440    • writing effective config file
#14 1.440       • writing                   config=/tmp/dist955478537/config.yaml
#14 1.440    • building binaries
#14 1.441       • building                  binary=/tmp/dist955478537/faq_darwin_amd64/faq
#14 7.293       • running hook              hook=cp "/tmp/dist955478537/faq_darwin_amd64/faq" "/usr/local/bin/faq"
#14 7.304    • archives
#14 7.305       • creating                  archive=/tmp/dist955478537/faq_0.0.7-next_darwin_amd64.tar.gz
#14 8.856    • homebrew tap formula     
#14 8.869       • guessing install to be "bin.install \"faq\""
#14 8.871       • writing                   formula=/tmp/dist955478537/faq.rb
#14 8.871    • storing artifact list    
#14 8.873       • writing                   file=/tmp/dist955478537/artifacts.json
#14 8.873    • release succeeded after 7.48s
#14 8.890 INF: /out/faq_0.0.7-next_darwin_amd64.tar.gz
#14 8.903 INF: checksum file created: /out/faq_0.0.7-next_darwin_amd64.tar.gz.sha256
```

There is a linker issue with `darwin/arm64` atm:

```
#17 1.283    • snapshotting
#17 1.283       • building snapshot...      version=0.0.7-next
#17 1.283    • checking distribution directory
#17 1.283    • loading go mod information
#17 1.288    • writing effective config file
#17 1.290       • writing                   config=/tmp/dist1461289652/config.yaml
#17 1.290    • building binaries
#17 1.291       • building                  binary=/tmp/dist1461289652/faq_darwin_arm64/faq
#17 2.459    ⨯ release failed after 1.21s error=failed to build for darwin_arm64: exit status 2: github.com/jzelinskie/faq/internal/jq
#17 2.459 # github.com/jzelinskie/faq/internal/jq
#17 2.459 ld: warning: ignoring file /osxcross/macports/pkgs/opt/local/lib/libjq.dylib, building for macOS-arm64 but attempting to link with file built for macOS-x86_64
#17 2.459 ld: warning: ignoring file /osxcross/macports/pkgs/opt/local/lib/libonig.dylib, building for macOS-arm64 but attempting to link with file built for macOS-x86_64
#17 2.459 Undefined symbols for architecture arm64:
#17 2.459   "_jq_compile_args", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jq_compile_args in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jq_compile_args)
#17 2.459   "_jq_format_error", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jq_format_error in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jq_format_error)
#17 2.459   "_jq_init", referenced from:
#17 2.459       __cgo_2b81333dd0fb_C2func_jq_init in _x003.o
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jq_init in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_C2func_jq_init, __cgo_2b81333dd0fb_Cfunc_jq_init )
#17 2.459   "_jq_next", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jq_next in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jq_next)
#17 2.459   "_jq_set_error_cb", referenced from:
#17 2.459       _gojq_set_error_cb in _x002.o
#17 2.459       _gojq_reset_error_cb in _x002.o
#17 2.459   "_jq_start", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jq_start in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jq_start)
#17 2.459   "_jq_teardown", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jq_teardown in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jq_teardown)
#17 2.459   "_jv_array_get", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_array_get in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_array_get)
#17 2.459   "_jv_array_length", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_array_length in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_array_length)
#17 2.459   "_jv_array_set", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_array_set in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_array_set)
#17 2.459   "_jv_array_sized", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_array_sized in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_array_sized)
#17 2.459   "_jv_copy", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_copy in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_copy)
#17 2.459   "_jv_dump_string", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_dump_string in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_dump_string)
#17 2.459   "_jv_false", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_false in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_false)
#17 2.459   "_jv_free", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_free in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_free)
#17 2.459   "_jv_get_kind", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_get_kind in _x003.o
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_is_valid in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_get_kind)
#17 2.459   "_jv_invalid_get_msg", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_invalid_get_msg in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_invalid_get_msg)
#17 2.459   "_jv_is_integer", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_is_integer in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_is_integer)
#17 2.459   "_jv_kind_name", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_kind_name in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_kind_name)
#17 2.459   "_jv_null", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_null in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_null)
#17 2.459   "_jv_number", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_number in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_number, __cgo_2b81333dd0fb_Cfunc_jv_number_value )
#17 2.459   "_jv_number_value", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_number_value in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_number_value)
#17 2.459   "_jv_object", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_object in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_object_set, __cgo_2b81333dd0fb_Cfunc_jv_object_iter_next , __cgo_2b81333dd0fb_Cfunc_jv_object_iter_key , __cgo_2b81333dd0fb_Cfunc_jv_object , __cgo_2b81333dd0fb_Cfunc_jv_object_iter_valid , __cgo_2b81333dd0fb_Cfunc_jv_object_iter , __cgo_2b81333dd0fb_Cfunc_jv_object_iter_value )
#17 2.459   "_jv_object_iter", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_object_iter in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_object_iter_next, __cgo_2b81333dd0fb_Cfunc_jv_object_iter_key , __cgo_2b81333dd0fb_Cfunc_jv_object_iter_valid , __cgo_2b81333dd0fb_Cfunc_jv_object_iter , __cgo_2b81333dd0fb_Cfunc_jv_object_iter_value )
#17 2.459   "_jv_object_iter_key", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_object_iter_key in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_object_iter_key)
#17 2.459   "_jv_object_iter_next", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_object_iter_next in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_object_iter_next)
#17 2.459   "_jv_object_iter_valid", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_object_iter_valid in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_object_iter_valid)
#17 2.459   "_jv_object_iter_value", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_object_iter_value in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_object_iter_value)
#17 2.459   "_jv_object_set", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_object_set in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_object_set)
#17 2.459   "_jv_parse", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_parse in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_parse)
#17 2.459   "_jv_string_sized", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_string_sized in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_string_sized)
#17 2.459   "_jv_string_value", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_string_value in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_string_value)
#17 2.459   "_jv_true", referenced from:
#17 2.459       __cgo_2b81333dd0fb_Cfunc_jv_true in _x003.o
#17 2.459      (maybe you meant: __cgo_2b81333dd0fb_Cfunc_jv_true)
#17 2.459 ld: symbol(s) not found for architecture arm64
#17 2.459 clang: error: linker command failed with exit code 1 (use -v to see invocation)
#17 2.459
#17 2.461 ERR: goreleaser failed: exit status 1
#17 ERROR: process "/bin/sh -c LDFLAGS=\"-s -w\"\nif [ \"$GO_LINKMODE\" = \"static\" ] && [ \"${TARGETOS}\" = \"linux\" ]; then\n  LDFLAGS=\"$LDFLAGS -linkmode external -extldflags -static\"\nfi\nif [ \"${TARGETOS}\" = \"darwin\" ]; then\n  export MACOSX_DEPLOYMENT_TARGET=10.10\n  echo 1 | osxcross-macports select-mirror\n  osxcross-macports install jq\n  export OSXCROSS_MP_INC=1\nfi\ngoreleaser-xx --debug \\\n  --config=\".goreleaser.yml\" \\\n  --name=\"faq\" \\\n  --dist=\"/out\" \\\n  --main=\"./cmd/faq\" \\\n  --flags=\"-v\" \\\n  --ldflags=\"$LDFLAGS -X 'github.com/jzelinskie/faq/pkg/version.Version={{.Version}}'\" \\\n  --tags=\"netgo\" \\\n  --files=\"LICENSE\" \\\n  --files=\"README.md\"\n" did not complete successfully: exit code: 1
```

Seems linked to https://github.com/tpoechtrager/osxcross/issues/313

You can try with `docker buildx bake --set *.platform=darwin/arm64`.

**Comment by crazy-max** [comment]
Looks like there is no MacPorts pkg targeting `arm64` available for `jq` atm:

```
#20 0.617  candidates for jq:
#20 0.617   jq-1.6_4.darwin_20.arm64.tbz2
#20 0.617   jq-1.6_4.darwin_20.arm64.tbz2
#20 0.617   jq-1.6_4.darwin_13.x86_64.tbz2
#20 0.617   jq-1.6_4.darwin_13.x86_64.tbz2
#20 0.617   jq-1.6_4.darwin_16.x86_64.tbz2
#20 0.617   jq-1.6_4.darwin_16.x86_64.tbz2
#20 0.617   jq-1.6_4.darwin_19.x86_64.tbz2
#20 0.617   jq-1.6_4.darwin_19.x86_64.tbz2
#20 0.617   jq-1.6_4.darwin_11.x86_64.tbz2
#20 0.617   jq-1.6_4.darwin_11.x86_64.tbz2
#20 0.618   jq-1.6_4.darwin_15.x86_64.tbz2
#20 0.618   jq-1.6_4.darwin_15.x86_64.tbz2
#20 0.618   jq-1.6_4.darwin_14.x86_64.tbz2
#20 0.618   jq-1.6_4.darwin_14.x86_64.tbz2
#20 0.618   jq-1.6_4.darwin_17.x86_64.tbz2
#20 0.618   jq-1.6_4.darwin_17.x86_64.tbz2
#20 0.618   jq-1.6_4.darwin_20.x86_64.tbz2
#20 0.618   jq-1.6_4.darwin_20.x86_64.tbz2
#20 0.618   jq-1.6_4.darwin_18.x86_64.tbz2
#20 0.618   jq-1.6_4.darwin_18.x86_64.tbz2
#20 0.618   jq-1.6_4.darwin_21.arm64.tbz2
#20 0.619   jq-1.6_4.darwin_21.arm64.tbz2
#20 0.619   jq-1.6_4.darwin_21.x86_64.tbz2
#20 0.619   jq-1.6_4.darwin_21.x86_64.tbz2
```

**Comment by crazy-max** [comment]
@jzelinskie Ok that's because osxcross doesn't support macports for `jq` against `arm64`. We need at least `darwin_21`. I will open a PR with https://github.com/crazy-max/osxcross/commit/aee7305d8bf66e494248f374efc9fdbed989c11a to add support for it.

In the meantime I have created a new osxcross image `crazymax/osxcross:11.3-test` that supports it. See https://github.com/jzelinskie/faq/pull/96/commits/406ce1b928d73fff31fc5402323b427fbced4edf. Seems to work fine:

```
#20 2.614    • snapshotting
#20 2.614       • building snapshot...      version=0.0.7-next
#20 2.614    • checking distribution directory
#20 2.614    • loading go mod information
#20 2.620    • writing effective config file
#20 2.621       • writing                   config=/tmp/dist2432223994/config.yaml
#20 2.621    • building binaries
#20 2.621       • building                  binary=/tmp/dist2432223994/faq_darwin_arm64/faq
#20 5.238       • running hook              hook=cp "/tmp/dist2432223994/faq_darwin_arm64/faq" "/usr/local/bin/faq"
#20 5.249    • archives
#20 5.250       • creating                  archive=/tmp/dist2432223994/faq_0.0.7-next_darwin_arm64.tar.gz
#20 6.751    • homebrew tap formula     
#20 6.765       • guessing install to be "bin.install \"faq\""
#20 6.768       • writing                   formula=/tmp/dist2432223994/faq.rb
#20 6.768    • storing artifact list    
#20 6.769       • writing                   file=/tmp/dist2432223994/artifacts.json
#20 6.769    • release succeeded after 4.19s
#20 6.795 INF: /out/faq_0.0.7-next_darwin_arm64.tar.gz
#20 6.810 INF: checksum file created: /out/faq_0.0.7-next_darwin_arm64.tar.gz.sha256
#20 DONE 6.9s
```

**Comment by jzelinskie** [comment]
This looks really good! Thanks again for all of it. Just looks like there's an issue with the linker finding oniguruma and libjq.

**Comment by crazy-max** [comment]
> This looks really good! Thanks again for all of it. Just looks like there's an issue with the linker finding oniguruma and libjq.

Yes the issue with missing lib is about a small linker issue I have fixed and will be released soon on [goxx repo](https://github.com/crazy-max/goxx).

I have also created a PR with an example using a snippet of your project if it's ok with you: https://github.com/crazy-max/goxx/pull/5. Relevant commit: https://github.com/crazy-max/goxx/pull/5/commits/40958da5106107b461efa21334971b49ac97d072

**Comment by crazy-max** [comment]
@jzelinskie Ok everything seems to work fine now: https://github.com/crazy-max/faq/releases/tag/0.0.9

File: .goreleaser.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    folder: Formula
+    homepage: https://github.com/jzelinskie/faq
+    description: Format Agnostic jQ
+    license: "Apache 2.0"
+    dependencies:
+      - name: jq
+    test: |
+      system "#{bin}/faq", "--version"
+    # FIXME: Skip upload for now. Needs to handle multiple packages.
+    skip_upload: true
```
Comment (crazy-max) [inline .goreleaser.yml:28]: I have started to add support for homebrew taps but it upload is skipped for now. We need to handle multiple packages outside the GoReleaser pipeline or maybe we could use an internal invocation. We can do that in a follow-up if that's ok for you.

File: docker-bake.hcl [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+target "test" {
+  target = "test-coverage"
+  output = ["."]
+}
+
+target "artifact" {
+  target = "artifacts"
+  secret = ["id=GITHUB_TOKEN,env=GITHUB_TOKEN"]
```
Comment (crazy-max) [inline docker-bake.hcl:26]: We can manage secrets at build time. Here the GITHUB_TOKEN we be used in the `artifact` stage in the Dockerfile.

File: Dockerfile [hunk 4]
```diff
@@ -1,10 +1,61 @@
...
+COPY --from=test /tmp/coverage.txt /coverage.txt
+
+FROM vendored AS build
+ARG TARGETPLATFORM
+ENV CGO_ENABLED=1
+ENV OSXCROSS_MP_INC=1
+RUN --mount=type=cache,sharing=private,target=/var/cache/apt \
+  --mount=type=cache,sharing=private,target=/var/lib/apt/lists \
+  goxx-apt-get install -y binutils gcc g++ pkg-config libjq-dev libonig-dev
+RUN goxx-macports install jq
```
Comment (crazy-max) [inline Dockerfile:41]: Here is the magic section in the Dockerfile to handle the specific toolchain packages based on the `TARGETPLATFORM` value. More info about the `goxx-*` wrappers: https://github.com/crazy-max/goxx#wrappers

Comment (crazy-max) [reply Dockerfile:41]: For windows I will work on a cygwin port wrapper. https://github.com/crazy-max/goxx/issues/14

**Comment by crazy-max** [comment]
@jzelinskie Looks good :tada:

image

[Artifacts available in the workflow](https://github.com/jzelinskie/faq/actions/runs/1645210840) if you want to check them: https://github.com/jzelinskie/faq/suites/4808306636/artifacts/135899023

**Comment by jzelinskie** [comment]
I'm still not sure the linking is correct for darwin or at least for homebrew users, which now have the new default of `/opt/homebrew` rather than in `/usr/local`.

```
$ uname -a
Darwin hanazawa.local 21.2.0 Darwin Kernel Version 21.2.0: Sun Nov 28 20:28:41 PST 2021; root:xnu-8019.61.5~1/RELEASE_ARM64_T6000 arm64
$ ./faq
dyld[39652]: Library not loaded: /opt/local/lib/libjq.1.dylib
  Referenced from: /Users/jzelinskie/Downloads/faq
  Reason: tried: '/opt/local/lib/libjq.1.dylib' (no such file), '/usr/local/lib/libjq.1.dylib' (no such file), '/usr/lib/libjq.1.dylib' (no such file)
[1]    39652 abort      ./faq
$ otool -L faq
faq:
	/opt/local/lib/libjq.1.dylib (compatibility version 2.0.0, current version 2.4.0)
	/opt/local/lib/libonig.5.dylib (compatibility version 8.0.0, current version 8.0.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1775.118.101)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
$ ls -l /opt/homebrew/include/jq.h
lrwxr-xr-x 29 jzelinskie 21 Dec  2021 /opt/homebrew/include/jq.h -> ../Cellar/jq/1.6/include/jq.h
```

**Comment by crazy-max** [comment]
@jzelinskie Oh good catch, was able to repro too:

```
dyld[51572]: Library not loaded: /opt/local/lib/libjq.1.dylib
  Referenced from: /private/tmp/faq
  Reason: tried: '/opt/local/lib/libjq.1.dylib' (no such file), '/usr/local/lib/libjq.1.dylib' (no such file), '/usr/lib/libjq.1.dylib' (no such file)
zsh: abort      ./faq
```

It's linked to MacPorts. Will fix that.

**Comment by crazy-max** [comment]
@jzelinskie Ok was able to fix that, missing the `--static` flag for the MacPorts command. Will push my changes:

```
crazymax@mac-mini-m1 dist % otool -L faq                                                                      
faq:
        /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1775.118.101)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
crazymax@mac-mini-m1 dist % ./faq 
Error: no arguments provided
Usage:
  faq [flags] [filter string] [files...]

Flags:
      --args string                  Takes a value and adds it to the position arguments list. Values are always strings. Positional arguments are available as $ARGS.positional[]. Specify --args multiple times to pass additional arguments. (default "[]")
  -C, --color-output                 colorize the output (default true)
  -c, --compact-output               compact output (don't pretty print the output)
  -h, --help                         help for faq
  -f, --input-format string          input format (default "auto")
      --jsonargs <jsonValue>         Takes a value and adds it to the position arguments list. Values are parsed as JSON values. Positional arguments are available as $ARGS.positional[]. Specify --jsonargs multiple times to pass additional arguments. (default null)
      --jsonkwargs key=<jsonValue>   Takes a key=value pair, setting $key to the JSON value of <value>: --kwargs foo={"fizz": "buzz"} sets $foo to the json object {"fizz": "buzz"}. Values are parsed as JSON values. Named arguments are also available as $ARGS.named[]. Specify --jsonkwargs multiple times to add more arguments. (default map[])
      --kwargs key=string            Takes a key=value pair, setting $key to <value>: --kwargs foo=bar sets $foo to "bar". Values are always strings. Named arguments are also available as $ARGS.named[]. Specify --kwargs multiple times to add more arguments. (default map[])
  -M, --monochrome-output            monochrome (don't colorize the output)
  -n, --null-input null              use null as the single input value
  -o, --output-format string         output format (default "auto")
  -p, --pretty-output                pretty-printed output (default true)
  -F, --program-file string          If specified, read the file provided as the jq program for faq.
  -r, --raw-output                   output raw strings, not JSON texts
  -s, --slurp                        read (slurp) all inputs into an array; apply filter to it
  -v, --version                      Print the version and exit.
```

**Review by jzelinskie — APPROVED** [review]
LGTM!
Thanks a ton for shepherding this change!

**Comment by crazy-max** [comment]
> LGTM! Thanks a ton for shepherding this change!

@jzelinskie You're welcome! Don't hesitate to ping me if you found some issues with this workflow.

**Comment by crazy-max** [comment]
Looks like the multiplatform image has been successfully pushed [on quay](https://quay.io/repository/jzelinskie/faq?tag=latest&tab=tags):

image

**Comment by jzelinskie** [comment]
Doesn't seem like the images work:

```sh
$ docker run -ti quay.io/jzelinskie/faq:0.0.8 --version
standard_init_linux.go:228: exec user process caused: no such file or directory
```

It also looks like homebrew didn't get updated: https://github.com/jzelinskie/homebrew-faq

**Comment by jzelinskie** [comment]
The downloaded images also still have `devel` as their version and aren't linked with their tag values:
```sh
$ ./faq --version
github.com/jzelinskie/faq/cmd/faq version (devel)
	github.com/Azure/draft v0.16.0
	github.com/BurntSushi/toml v0.3.1
	github.com/alecthomas/chroma v0.8.2
	github.com/clbanning/mxj/v2 v2.5.5
	github.com/danwakefield/fnmatch v0.0.0-20160403171240-cbb64ac3d964
	github.com/dlclark/regexp2 v1.2.0
	github.com/ghodss/yaml v1.0.0
	github.com/globalsign/mgo v0.0.0-20181015135952-eeefdecb41b8
	github.com/go-xmlfmt/xmlfmt v0.0.0-20191208150333-d5b6f63a941b
	github.com/jbrukh/bayesian v0.0.0-20200318221351-d726b684ca4a
	github.com/sirupsen/logrus v1.8.0
	github.com/spf13/cobra v1.1.3
	github.com/spf13/pflag v1.0.5
	github.com/zeebo/bencode v1.0.0
	golang.org/x/crypto v0.0.0-20210220033148-5ea612d1eb83
	golang.org/x/net v0.0.0-20190620200207-3b0461eec859
	golang.org/x/sys v0.0.0-20200413165638-669c56c373c4
	golang.org/x/term v0.0.0-20201117132131-f5c789dd3221
	golang.org/x/text v0.3.2
	gopkg.in/yaml.v2 v2.4.0
	howett.net/plist v0.0.0-20201203080718-1454fab16a06
```

**Comment by crazy-max** [comment]
@jzelinskie 

> Doesn't seem like the images work:
> 
> ```shell
> $ docker run -ti quay.io/jzelinskie/faq:0.0.8 --version
> standard_init_linux.go:228: exec user process caused: no such file or directory
> ```

Yes it's a scratch image and linux bins are not statically linked. Will fix that.

> It also looks like homebrew didn't get updated: https://github.com/jzelinskie/homebrew-faq

Yes atm [upload is skipped](https://github.com/jzelinskie/faq/blob/main/.goreleaser.yml#L27). You can update manually the formula on your own for now. Will open a follow-up PR to handle that.

> The downloaded images also still have `devel` as their version and aren't linked with their tag values:

This git status seems ok: https://github.com/jzelinskie/faq/runs/4749795960?check_suite_focus=true#step:7:845

```
#43 0.751 INF: git status:
#43 0.752   tag=0.0.8
#43 0.752   dirty=false
#43 0.753   wrongref=false
```

So I guess there is something wrong with the ldflags: https://github.com/jzelinskie/faq/blob/ef38339327eb46352d3c7c317331b8fd5249b249/Dockerfile#L54. Will take a look on that too.

**Comment by crazy-max** [comment]
> So I guess there is something wrong with the ldflags: 

Got it, it should be `github.com/jzelinskie/faq/internal/version.`

**Pull request merged**